crates/pathfinder/src/state.rs

```diff
@@ -1,80 +1,438 @@
-//! Contains the [GlobalStateTree] and [ContractsStateTree] trees, which combined
-//! store the total StarkNet state.
-//!
-//! These are abstractions built-on the [Binary Merkle-Patricia Tree](MerkleTree).
-
-use rusqlite::Transaction;
+use anyhow::Context;
+use pedersen::{pedersen_hash, StarkHash};
+use rusqlite::{Connection, Transaction};
+use web3::{types::H256, Transport, Web3};
 
 use crate::{
     core::{
-        ContractAddress, ContractRoot, ContractStateHash, GlobalRoot, StorageAddress, StorageValue,
+        ContractHash, ContractRoot, ContractStateHash, GlobalRoot, StarknetBlockHash,
+        StarknetBlockTimestamp,
+    },
+    ethereum::{
+        log::{FetchError, StateUpdateLog},
+        state_update::{
+            state_root::StateRootFetcher, ContractUpdate, DeployedContract,
+            RetrieveStateUpdateError, StateUpdate,
+        },
+        BlockOrigin, EthOrigin, TransactionOrigin,
     },
-    state::merkle_tree::MerkleTree,
+    rpc::types::{BlockHashOrTag, BlockNumberOrTag, Tag},
+    sequencer,
+    state::state_tree::{ContractsStateTree, GlobalStateTree},
+    storage::{ContractsStateTable, ContractsTable, GlobalStateTable},
 };
 
 mod merkle_node;
 mod merkle_tree;
+mod state_tree;
+
+#[derive(thiserror::Error, Debug)]
+enum UpdateError {
+    #[error("Ethereum chain reorg detected")]
+    Reorg,
+    #[error(transparent)]
+    Other(#[from] anyhow::Error),
+}
+
+/// Syncs the Starknet state with L1.
+pub async fn sync<T: Transport>(
+    database: &mut Connection,
+    transport: &Web3<T>,
+    sequencer: &sequencer::Client,
+) -> anyhow::Result<()> {
+    // TODO: Track sync progress in some global way, so that RPC can check and react accordingly.
+    //       This could either be the database, or a mutable lazy_static thingy.
+
+    let db_tx = database
+        .transaction()
+        .context("Create database transaction")?;
+
+    let latest_state =
+        GlobalStateTable::get_latest_state(&db_tx).context("Get latest StarkNet state")?;
+
+    // There shouldn't be any side effects from just reading the state,
+    // but rolling back to be safe.
+    //
+    // We don't care about any errors from rolling back the transaction.
+    let _ = db_tx.rollback();
```

Comment:
I don't understand the rationale here ... Okay, so maybe you just want to quickly read the current latest, trust that it's the same when you start a transaction again.

This might be more intention revealing:

```
let latest_state = {
    let tx = database.transaction()?;
    let st = GlobalStateTable::...;
    let _ = tx.rollback();
    st
};
```

as I was looking for if this is reused later on.

However I think this could be done at the near-beginning of the loop as well, or at minimum verified that this is indeed the latest.

Reply:
Discussed offline; summary:

Rather make no assumptions about this code being the only writer, but instead validate this by checking latest state from database for every update.